### PR TITLE
fallback to use copy if rename fails

### PR DESCRIPTION
**Pull request by rajatjindal**
I am trying to add statically linked spin binary creation as part of the CI (this is required for our openshift integration effort). 

As part of this, I am using `cross` to build for `x86_64-unknown-linux-musl` target, which apparently runs inside a docker container. 

While doing that, I ran into an error as follows:

```
  Finished release [optimized + debuginfo] target(s) in 35.29s
  thread 'main' panicked at /cargo/git/checkouts/cargo-target-dep-97a1ed8d29020898/b7b1989/src/lib.rs:111:17:
  Failed to move output "/target/x86_64-unknown-linux-musl/release/build/spin-cli-76b8343b474a1d22/out/target-deps/target__test-programs__core-wasi-test.wasm/wasm32-wasi/release/core-wasi-test.wasm" to "target/test-programs/core-wasi-test.wasm": Invalid cross-device link (os error 18)
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

using copy/removing the artifact seems to have fixed this. 

The current change assumes it is a file that we are renaming. do you think we need to handle dir scenario as well.

Also, it seems like https://github.com/moby/moby/pull/43225 was fixed and made available in 24.0.0 version of moby, but it is not available yet in docker (https://github.com/docker/cli/blob/fefadb9be671251e92c410830628ba4188d6b974/vendor.mod#L15)

### CODE REVIEW

**Comment by lann** [comment]
Could this _attempt_ the rename first and then fallback to copy if that fails? Renaming is much faster in the common case where it works.

File: src/lib.rs [hunk 1]
```diff
@@ -107,11 +109,12 @@ impl TargetDep {
...
-                panic!(
-                    "Failed to move output {:?} to {:?}: {}",
-                    out_path, &self.output_path, err
-                )
+            std::fs::copy(out_path, &self.output_path).unwrap_or_else(|err| {
+                panic!("error when copying file: {:?}", err)
+            });
+
+            std::fs::remove_file(out_path).unwrap_or_else(|err| {
+                panic!("error when removing file: {:?}", err)
```
Comment (lann) [inline src/lib.rs]: Probably not important to remove. The rename here is just for performance, not because it needs "move" semantics.

Comment (rajatjindal) [reply src/lib.rs]: removed the `remove_file` logic

File: src/lib.rs [hunk 2]
```diff
@@ -108,12 +110,16 @@ impl TargetDep {
 
             // Move output
             std::fs::rename(out_path, &self.output_path).unwrap_or_else(|err| {
-                panic!(
-                    "Failed to move output {:?} to {:?}: {}",
+                println!(
```
Comment (lann) [inline src/lib.rs]: `stderr` is safer for logging; `stdout` can carry build metadata in cargo build scripts (see line 125 below)
```suggestion
                eprintln!(
```

File: src/lib.rs [hunk 3]
```diff
@@ -108,12 +110,16 @@ impl TargetDep {
 
             // Move output
             std::fs::rename(out_path, &self.output_path).unwrap_or_else(|err| {
```
Comment (lann) [inline src/lib.rs]: For this kind of thing `if let Err(err) = ...` is a little easier to follow than `unwrap_or_else`.

**Review by lann — COMMENTED** [review]
Logging to stderr is blocking; other suggestion isnt

File: src/lib.rs (hunk 2 above)
Comment (rajatjindal) [reply src/lib.rs]: done. thank you.

File: src/lib.rs (hunk 3 above)
Comment (rajatjindal) [reply src/lib.rs]: thank you. changed this as well.

**Review by lann — APPROVED** [review]
Thanks!

**Pull request merged**